File: Java/Cryptography/Affine_Cipher.java
//AFFINE CIPHER
//The Affine cipher is a standard substitution cipher.
//Encryption:
//        Cipher Text, C=(aX+b) mod 26
//        where a, b are key values and X is letter to be encrypted.
//Decryption:
//        Plain Text, X=[a^-1(C-b)] mod 26
//        where alphabets from (a-z) are numbered sequentially from (0-25).

import java.math.BigInteger;
import java.util.Scanner;

public class Affine_Cipher{
    static int a, b; //Keys for affine cipher
    static int m = 26;//total number of alphabets

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter plain text:");
        String plainText = sc.nextLine();

        System.out.println("Enter keys a and b:");
        a = sc.nextInt();
        b = sc.nextInt();

        if (!isCoprime(a, m)) {
            System.out.println("Key 'a' and modulus 'm' are not coprime. Choose different keys.");
            return;
        }

        StringBuilder ciphertext = cipher(plainText);
        System.out.println("Cipher Text after Encryption:");
        System.out.println(ciphertext);

        StringBuilder decryptedText = decipher(ciphertext);
        System.out.println("Decrypted Text/Plain Text after Decryption:");
        System.out.println(decryptedText);
    }

    static StringBuilder cipher(String plain_text) {
        StringBuilder ciphertext = new StringBuilder();
        for (int i = 0; i < plain_text.length(); i++) {
            char t = plain_text.charAt(i);
            int alphaVal = (int) t - 97; // 'a' has ASCII value 97
            int c = (a * alphaVal + b) % m;
            char changedChar = (char) (c + 97);
            ciphertext.append(changedChar);
        }
        return ciphertext;
    }

    static StringBuilder decipher(StringBuilder ciphertext) {
        StringBuilder plaintext = new StringBuilder();
        BigInteger inverseA = BigInteger.valueOf(modInverse(a, m));

        for (int i = 0; i < ciphertext.length(); i++) {
            char c = ciphertext.charAt(i);
            int alphaVal = (int) c - 97;
            int decipherT = (inverseA.intValue() * (alphaVal - b + m)) % m;
            char changedChar = (char) (decipherT + 97);
            plaintext.append(changedChar);
        }
        return plaintext;
    }

    static boolean isCoprime(int a, int b) {
        return BigInteger.valueOf(a).gcd(BigInteger.valueOf(b)).intValue() == 1;
    }

    static int modInverse(int a, int m) {
        BigInteger bigA = BigInteger.valueOf(a);
        BigInteger bigM = BigInteger.valueOf(m);
        return bigA.modInverse(bigM).intValue();
    }
}

File: Python/Cryptography/Affine_Cipher.py
#AFFINE CIPHER
#The Affine cipher is a standard substitution cipher.
#Encryption:
#        Cipher Text, C=(aX+b) mod 26
#        where a, b are key values and X is letter to be encrypted.
#Decryption:
#        Plain Text, X=[a^-1(C-b)] mod 26
#        where alphabets from (a-z) are numbered sequentially from (0-25).

def is_coprime(a, b):
    while b != 0:
        temp = b
        b = a % b
        a = temp
    gcd = a
    return gcd == 1

def mod_inverse(a, m):
    a = a % m
    for x in range(1, m):
        if (a * x) % m == 1:
            return x
    return -1

def cipher(plain_text, a, b):
    ciphertext = ""
    for t in plain_text:
        alpha_val = ord(t) - 97
        c = (a * alpha_val + b) % 26
        changed_char = chr(c + 97)
        ciphertext += changed_char
    return ciphertext

def decipher(ciphertext, a, b):
    plaintext = ""
    inverse_a = mod_inverse(a, 26)
    for c in ciphertext:
        alpha_val = ord(c) - 97
        decipher_t = (inverse_a * (alpha_val - b + 26)) % 26
        changed_char = chr(decipher_t + 97)
        plaintext += changed_char
    return plaintext

def main():
    plain_text = input("Enter plain text: ")
    a, b = map(int, input("Enter keys a and b: ").split())

    if not is_coprime(a, 26):
        print("Key 'a' and modulus 'm' are not coprime. Choose different keys.")
        return

    ciphertext = cipher(plain_text, a, b)
    print("Cipher Text after Encryption:")
    print(ciphertext)

    decrypted_text = decipher(ciphertext, a, b)
    print("Decrypted Text/Plain Text after Decryption:")
    print(decrypted_text)

if __name__ == "__main__":
    main()
